src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,670 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using RoslynTrigger = Microsoft.CodeAnalysis.Completion.CompletionTrigger;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal class EditorAsyncCompletionItemManager : IAsyncCompletionItemManager
+    {
+        private readonly IAsyncCompletionBroker _broker;
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        public EditorAsyncCompletionItemManager(IAsyncCompletionBroker broker)
+        {
+            _broker = broker;
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        // This method is supposed to be called just once per session. That is why we register events there.
+        // If the idea is changed, we need an Initialize method to be added to the interface.
```

Comment:
This seems at odds with the documentation for the method:

> This method is first called before completion is about to appear, and *then on subsequent typing events* and when user toggles completion filters.


Why can't the editor just add an IAsyncCompletionItemManager.OnItemCommitted() method they call back with?

Reply:
the comment is incorrect. I'm going to fix it. Thank you for pointing this out!